### DISCUSSION TITLE
Connected endpoints and Tenants

### DISCUSSION THREAD

**Discussion by PieterL75** [aristanetworks/avd#4835] [Q&A]
Is there a way to link 'endpoint_ports' to a Tenant ?
I have 2 tenants, with overlapping VLAN ids, but unique VNI's and Vlan-aware-bundles.

When setting a vlan that exists in both tenants to a endpoint_port, I get an error that is is duplicate.

Is there a way to define what Tenant the vlan has to be taken from ?
I do need both Tenants on that leaf, but just a few non-overlapping vlans of both

**Comment by ClausHolbechArista** [comment]
The only way to handle this would be by using `tags` under the SVIs/L2VLANs and `tags` under the `filter` on the node-config. That way you can further limit the "scope" of VLANs that are usable by this device. I believe it should also work in conjuction with `only_vlans_in_use` if you use that.

**Reply by PieterL75** [reply]
I have only_vlans_in_use enabled indeed.. but it would make a mgmt nightmare to tag the vlans for each device/pair they should exists on...

**Reply by ClausHolbechArista** [reply]
Then I suggest you split the tenants for each group of "overlapping vlans"

**Reply by ClausHolbechArista** [reply]
We have talked about having a new optional "vlan_number" key per vlan, so the ID can be truly unique, and this way we translate the vlan when we generate. The ID would then still be used on the connected endpoints. But it quickly becomes a nightmare to manage as well, so we have not moved on with this.

Suggestions are most welcome.

**Reply by PieterL75** [reply]
What if we use the VNI in the 'vlan' list? that can be translated back to a VLAN.
Then you can even use 2 vlans of different tenants on one uplink.

Overlapping vlans are ofcourse not possible on one device, if they are used

**Reply by ClausHolbechArista** [reply]
Could be one way. Since VNIs are generated based on the tenant base VNI, it could also become very hard to decipher.
Let me discuss a bit with the team, but please add more ideas if you think of something. I don't think we can "bind" an endpoint/network_port to a single tenant to resolve this.

**Reply by PieterL75** [reply]
Indeed, one endpoint could contain vlans of different tenants...
another option is to add the tenantname to the vlan if there is a duplicate vlan id across the assigned tenants to the devices.
if vlan 100 and 101 is used in 2 tenants, and vlan 102 only in one, then you could so something like
  `vlans 100[tenant1], 101[tenant3], 102`
(I'm talking in the case of `only_vlans_in_use ` being enabled

For the VNI's: we use a very simple schema 50xxxx, 10xxxx; with the tenant being 10,50 and xxxx the vlan.. So for us that could be doable. But I can imagine, that if you have a lot more tenants it could be more difficult.

**Reply by PieterL75** [reply]
Or foresee a new `VNIs:` field under the endpoint_ports.
the `vlans:` will still look for the unique vlans across tenants, while te `vni:` is used for overlapping.. so both statements should cooexist

**Reply by PieterL75** [reply]
Idea of this night, a vlan could also include it's vni.
Ex:
vlans: 100,101@12340101,102@45670102,103
I used an @, but that's just an example

**Comment by PieterL75** [comment]
I've been looking into the code, but it looks like a difficult one..
all vlans are added to one dict, with the vlan ID as the 'unique' identifier..

I guess it needs some other mean to add vlans to the dict, and set the primary key to the "vlan_vni" or "vni or vlan"

**Comment by PieterL75** [comment]
@ClausHolbechArista Is there an option in the new releases that could cover this ?
I have several tenants defined, and some of them have an overlapping VLAN.
When I a use a 'filter/tenants' on my device group, and then use a VLAN that exists in both tenants, I get an overlapping error.
Is there a way to specifically specify a vlan from a tenant into a portgroup ?

**Reply by ClausHolbechArista** [reply]
The only way we have Today would be to use tags under the vlans and on the device filters. That way the available vlans will only be what matches the tags and you can avoid the collisions.
